### PR TITLE
Improved log interface performance and streaming

### LINKED ISSUE #25: Improve Log View/Parsing Performance
Both container infrastructures make some slow operations (also noted in #24). 

- String/log line parsing ([potential optimizations](https://lise-henry.github.io/articles/optimising_strings.html))
- Reliance on APIs that do not support head

We should improve PREvant's handling of log APIs

### PR DESCRIPTION
**Pull request by samuchila**
This commit enhances the log interface by reducing lag and improving the rendering speed. It introduces a feature to load the latest logs first and logs are now streamed in real-time from backend, ensuring immediate access to new logs. The infrastructure is now able to handle both streaming and one time log transfer in one function only, avoiding the use of multiple implementations. The routes ensure the output is served to the client based on content negotiation.

Fixes #25

### CODE REVIEW

**Comment by samuchila** [comment]
@schrieveslaach Please review the changes.

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
Thanks for tackling this issue. I think it is a good start but in the details I want to iron out somethings:

- `tail_lines` seems to mixed up a little
- It seems to me that we don't need two implementations under the hood

Let's have a chat to figure out these things.

File: api/src/apps/routes.rs [hunk 1]
```diff
@@ -211,7 +216,7 @@ async fn logs(
             }
         },
     };
-    let limit = limit.unwrap_or(20_000);
+    let limit = limit.unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs:219]: I understand the mistake, maintaining limit as an `Option` at the infrastructure level instead of unwrapping in `routes` will result in better handling.

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
Architecture has been improved significantly. Love it. 

There are some minor details (see comment). Also, the frontend could be simplified when downloading logs.

File: api/src/apps/mod.rs [hunk 2]
```diff
@@ -703,6 +721,40 @@ Log msg 3 of service-a of app master
...
+                    assert_eq!(
+                        log_line,
+                        format!(
+                            "Log msg {log_count} of {} of app {app_name}\n",
+                            service.service_name()
+                        )
+                    );
+                    log_count += 1;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/mod.rs:773]: No real value, just an aid for the loop. The main point here is to test the `streaming` mechanism rather than the count. Have updated it.

File: api/src/apps/routes.rs [hunk 3]
```diff
@@ -211,20 +215,56 @@ async fn logs(
...
+    limit: Option<usize>,
+    since: Option<String>,
+    apps: &'a State<Arc<Apps>>,
+) -> EventStream![Event + 'a] {
+    let app_name = app_name.unwrap();
+    let since = match since {
+        None => None,
+        Some(since) => match DateTime::parse_from_rfc3339(&since) {
+            Ok(since) => Some(since),
+            Err(_) => None,
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs]: Agreed, updated as suggested.

File: api/src/apps/routes.rs [hunk 4]
```diff
@@ -211,20 +215,56 @@ async fn logs(
...
+        let mut log_chunk =
+            apps
+                .stream_logs(&app_name, service_name, &since, &limit)
+                .await;
+        while let Some(result) = log_chunk.as_mut().next().await {
+            match result {
+                Ok((_,log_line)) => {
+                    yield Event::data(log_line)
+                },
+                Err(_e) => { break;}
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs]: I observed this too. Prompted me to take another look - seems culprit is that rustfmt does not format anything in between a `macro with a flower bracket`.  Refer for a long read - https://github.com/rust-lang/rustfmt/issues/8
TLDR, Had to format the code outside the macro and place it inside the macro.

File: api/src/apps/routes.rs [hunk 5]
```diff
@@ -211,20 +215,56 @@ async fn logs(
             }
         },
     };
-    let limit = limit.unwrap_or(20_000);
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs:214]: Updated.

File: api/src/apps/routes.rs [hunk 6]
```diff
@@ -300,15 +340,23 @@ impl<'r> Responder<'r, 'static> for LogsResponse {
...
+        let next_logs_url = match self.limit {
+            Some(limit) => format!(
+                "/api/apps/{}/logs/{}?limit={}&since={}",
+                self.app_name,
+                self.service_name,
+                limit,
+                RawStr::new(&from.to_rfc3339()).percent_encode(),
+            ),
+            None => format!(
+                "/api/apps/{}/logs/{}?since={}",
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs:372]: The next link without a limit will be used to start the stream, which does not need a limit.
Added tests to verify this.

File: frontend/src/LogsDialog.vue [hunk 7]
```diff
@@ -40,119 +52,187 @@
...
             this.$router.push('/');
          },
 
-         updateLogs() {
-            if ( this.nextPageLink ) {
-               const nextPageLink = this.nextPageLink;
-               this.nextPageLink = null;
-               this.currentPageLink = nextPageLink;
+         processDownload() {
+            fetch(this.downloadLink)
```
[the remark this answers is not in the document]
Comment (samuchila) [reply frontend/src/LogsDialog.vue]: Involved redesigning the api only slightly, but this got rid of the download function itself.

File: api/res/openapi.yml [hunk 8]
```diff
@@ -246,7 +246,7 @@ paths:
           example: '2019-07-22T08:42:47-00:00'
         - in: query
           name: limit
-          description: The number of log lines to retrieve. If not present, 1000 lines will be retrieved.
+          description: The number of log lines to retrieve. If not present, all the lines from `since` are retrieved.
```
Comment (schrieveslaach) [inline api/res/openapi.yml:249]: The new query parameter and the new content-type is not yet included here.

File: api/src/apps/routes.rs [hunk 9]
```diff
@@ -186,14 +190,16 @@ async fn change_status(
 }
 
 #[get(
-    "/<app_name>/logs/<service_name>?<since>&<limit>",
-    format = "text/plain"
+    "/<app_name>/logs/<service_name>?<since>&<limit>&<download>",
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:193]: I struggle a bit with the term `download`. What do you think about following suggestion?

```suggestion
    "/<app_name>/logs/<service_name>?<since>&<limit>&<asAttachment>",
```

File: api/src/apps/routes.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (schrieveslaach) [reply api/src/apps/routes.rs:364]: yeah, make sense. Sorry about providing confusion.

File: api/src/apps/routes.rs [hunk 10]
```diff
@@ -300,19 +350,39 @@ impl<'r> Responder<'r, 'static> for LogsResponse<'r> {
...
-        );
+        let content_dispositon_value = if self.download {
+            format!(
+                "attachment; filename=\"{}_{}_{}.txt\"",
+                self.app_name,
+                self.service_name,
+                Utc::now().format("%Y%m%d_%H%M%S")
+            )
+        } else {
+            format!("inline")
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:379]: No need for formatting anything.

```suggestion
            String::from("inline")
```

File: api/src/apps/routes.rs [hunk 11]
```diff
@@ -300,19 +350,39 @@ impl<'r> Responder<'r, 'static> for LogsResponse<'r> {
...
-            self.service_name,
-            self.limit,
-            RawStr::new(&from.to_rfc3339()).percent_encode(),
-        );
+        let content_dispositon_value = if self.download {
+            format!(
+                "attachment; filename=\"{}_{}_{}.txt\"",
+                self.app_name,
+                self.service_name,
+                Utc::now().format("%Y%m%d_%H%M%S")
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:376]: I think there shouldn't be use of `Utc::now()` and it should be derived from the log data.

```suggestion
                *log_chunk.until().format("%Y%m%d_%H%M%S")
```

File: api/src/apps/routes.rs [hunk 12]
```diff
@@ -766,6 +836,48 @@ mod tests {
...
+        #[tokio::test]
+        async fn render_log_weblink_with_no_limit() -> Result<(), crate::apps::AppsServiceError> {
+            let (host_meta_cache, mut _host_meta_crawler) = crate::host_meta_crawling();
+
+            let client =
+                set_up_rocket_with_dummy_infrastructure_and_a_running_app(host_meta_cache).await?;
+
+            let response = client
+                .get("/api/apps/master/logs/service-a")
+                .header(ContentType::Text)
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:849]: Wrong header here in the context of a client. [A client sends the accept header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept):

> The Accept request HTTP header indicates which content types, expressed as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types), the client is able to understand. The server uses [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) to select one of the proposals and informs the client of the choice with the [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) response header. 

When you change it to the Accept header it should fail because you also modified the content-type in the production code.

File: api/src/apps/routes.rs [hunk 13]
```diff
@@ -766,6 +836,48 @@ mod tests {
...
+        #[tokio::test]
+        async fn render_log_weblink_with_some_limit() -> Result<(), crate::apps::AppsServiceError> {
+            let (host_meta_cache, mut _host_meta_crawler) = crate::host_meta_crawling();
+
+            let client =
+                set_up_rocket_with_dummy_infrastructure_and_a_running_app(host_meta_cache).await?;
+
+            let response = client
+                .get("/api/apps/master/logs/service-a?limit=20000&since=2019-07-22T08:42:47-00:00")
+                .header(ContentType::Text)
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:871]: See comment above, wrong header.

File: frontend/src/LogsDialog.vue [hunk 14]
```diff
@@ -40,119 +52,166 @@
...
    let requestUri;
+   const itemSize = 24;
 
    export default {
       data() {
          return {
             logLines: [],
-            nextPageLink: null
+            eventSource: null,
+            scrollPosition: false,
```
Comment (schrieveslaach) [inline frontend/src/LogsDialog.vue:88]: Isn't following better?

```suggestion
            scrollPosition: null,
```

File: frontend/src/LogsDialog.vue [hunk 15]
```diff
@@ -24,11 +24,23 @@
 * =========================LICENSE_END==================================
 */
 <template>
-   <dlg ref="dialog" :title="`Logs of ${$route.params.service} in ${$route.params.app}`" :large="true" @close="clearLogs">
+   <dlg ref="dialog" :title="`Logs of ${$route.params.service} in ${$route.params.app}`" :large="true" @close="clearLogs"
+      :errorStatusText="this.errorText">
       <template v-slot:body>
-         <DynamicScroller :items="logLines" :item-size="20" class="ra-logs">
+         <div class="d-flex justify-content-end align-items-center">
+            <alert type="alert" class="alert-primary mr-auto" v-if="scrollPosition" role="alert">
```
Comment (schrieveslaach) [inline frontend/src/LogsDialog.vue:31]: Isn't following better because more robust?

```suggestion
            <alert type="alert" class="alert-primary mr-auto" v-if="scrollPosition === 0" role="alert">
```

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
Good work and I think there are just some minor things left. Please, checkout my comments.

File: api/src/apps/routes.rs (hunk 9 above)
Comment (samuchila) [reply api/src/apps/routes.rs:193]: camelCase throws a cargo warning. Perhaps you referred to using `as_attachment`

File: api/src/apps/routes.rs [hunk 16]
```diff
@@ -300,19 +350,39 @@ impl<'r> Responder<'r, 'static> for LogsResponse<'r> {
             Some(log_chunk) => log_chunk,
         };
 
-        let from = *log_chunk.until() + chrono::Duration::milliseconds(1);
+        let from = *log_chunk.until() + chrono::Duration::seconds(1);
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs:353]: Kubernetes supports only RFC3339 and Docker supports both RFC3339 and nano. The result for Kubernetes was that the millisecond change is ignored and during one of the tests a line or two of overlap was observed.

File: api/src/apps/routes.rs [hunk 17]
```diff
@@ -367,6 +379,7 @@ impl<'r> Responder<'r, 'static> for LogsResponse<'r> {
         let log_lines = log_chunk.log_lines();
         Response::build()
             .raw_header("Link", format!("<{}>;rel=next", next_logs_url))
+            .raw_header("Content-Disposition", self.content_disposition)
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs]: Yeah I thought about this, have made the change now.
However the design feels a bit off, the constant will still need to be converted to a string using `.as_str()` for both raw_header and Header::new combinations.
Looks something like this:
`.raw_header(rocket::http::hyper::header::CONTENT_DISPOSITION.as_str(), self.content_disposition)`

File: api/src/apps/routes.rs [hunk 18]
```diff
@@ -189,15 +189,16 @@ async fn change_status(
...
+    "/<app_name>/logs/<service_name>?<since>&<limit>&<download>",
+    format = "text/html",
     rank = 2
 )]
 async fn logs<'r>(
     app_name: Result<AppName, AppNameError>,
     service_name: &'r str,
     since: Option<String>,
     limit: Option<usize>,
+    download: bool,
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/apps/routes.rs:202]: I have not explicitly tested but I operated based on the documentation of Rocket. I will write a test however now.
Ref - https://rocket.rs/v0.5/guide/requests/#defaults

File: api/res/openapi.yml (hunk 8 above)
Comment (samuchila) [reply api/res/openapi.yml:249]: Offering the event-stream as it is in openapi.yml is not straightforward. There are a few ways to do it, I will confirm the applicable one and do it in the next hour.

**Review by schrieveslaach — APPROVED** [review]
Very, very nice.
